### PR TITLE
[10.0][ADD]stock_force_mto_rule

### PR DESCRIPTION
**Pull request by MPoch-PlanetaTIC**
Allows to force MTO rule in sale order lines

cc @PlanetaTIC

### CODE REVIEW

File: stock_force_mto_rule/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+# Copyright 2018 PlanetaTIC - Francisco Fernández <ffernandez@planetatic.com>
+# Copyright 2018 PlanetaTIC - Lluís Rovira <lrovira@planetatic.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    'name': 'Stock force MTO rule',
+    'version': '10.0.1.0.0',
+    'category': 'Warehouse',
+    'author': 'PlanetaTIC',
+    'website': 'https://www.planetatic.com',
```
Comment (rousseldenis) [inline stock_force_mto_rule/__manifest__.py]: Please change this according to https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst
 
and specifically https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py

File: stock_force_mto_rule/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    'description': "Allows to force MTO rule in sale order lines",
+    'depends': [
+        'sale',
+        'stock',
+    ],
+    'init_xml': [],
```
Comment (rousseldenis) [inline stock_force_mto_rule/__manifest__.py]: Please remove void entries.

File: stock_force_mto_rule/models/product.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    @api.model
+    def _get_mto_route(self):
+        mto_route_id = self.env['ir.model.data'].xmlid_to_res_id(
+            'stock.route_warehouse0_mto')
+        if mto_route_id:
+            return [mto_route_id]
```
Comment (rousseldenis) [inline stock_force_mto_rule/models/product.py]: Why not returning a recrodset instead ?

File: stock_force_mto_rule/models/sale.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    @api.multi
+    def _get_route_mto(self):
+        mto_route_id = self.env['product.template']._get_mto_route()
+        mto_route_id = mto_route_id and mto_route_id[0] or False
+        for line in self:
+            if mto_route_id and line.id:
+                line.route_mto = bool(self.search([
```
Comment (rousseldenis) [inline stock_force_mto_rule/models/sale.py]: Please do the search before the loop.

File: stock_force_mto_rule/models/sale.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            else:
+                line.route_mto = False
+
+    @api.multi
+    def _set_route_mto(self):
+        for line in self:
+            route_id = False
+            if line.route_mto:
+                if line.product_id:
+                    mto_route_ids = self.env['product.template'
```
Comment (rousseldenis) [inline stock_force_mto_rule/models/sale.py]: As this _get_mto_route() is an api.model, do it before the loop in a variable and reuse it.

File: stock_force_mto_rule/models/sale.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                                             ]._get_mto_route()
+                    if mto_route_ids:
+                        route_id = mto_route_ids[0]
+            line.route_id = route_id
+
+    route_mto = fields.Boolean(
```
Comment (rousseldenis) [inline stock_force_mto_rule/models/sale.py]: Please put fields on top of the file

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review. Some comments particular to performances.

Could you please add tests?

**Comment by MPoch-PlanetaTIC** [comment]
> Code review. Some comments particular to performances.
> 
> Could you please add tests?

Thank very much for your review.

Next week I will work on all requested points.

**Comment by MPoch-PlanetaTIC** [comment]
> Code review. Some comments particular to performances.
> 
> Could you please add tests?

Code reviewed.
I have added  a test and improved the performance.

Thank you.

**Review by rousseldenis — APPROVED** [review]
Code review

**Comment by rousseldenis** [comment]
@MPoch-PlanetaTIC Can you try to rebase ? If this fails and if no more needed, we could close it

**Comment by MPoch-PlanetaTIC** [comment]
> @MPoch-PlanetaTIC Can you try to rebase ? If this fails and if no more needed, we could close it

Right now, none of our clients work with Odoo 10. And if we have not migrated this module to Odoo 12, this means we do not use this module anymore.

You can close it.

Thank you

**Pull request closed without merging**